src/linear_algebra/alternating.lean

```diff
@@ -37,6 +43,57 @@ using `map_swap` as a definition, and does not require `has_neg N`.
 * `alternating_map.coe_smul`
 -/
 
+-- TODO: move
+section to_move
+
+namespace tactic
+namespace interactive
+
+open lean.parser
+open interactive
+
+/-- Focus on the first `n` goals. -/
+meta def first_n_goals : parse small_nat → itactic → tactic unit
+| n t := do
+  goals ← get_goals,
+  let current_goals := goals.take n,
+  let later_goals := goals.drop n,
+  set_goals current_goals,
+  t,
+  new_goals ← get_goals,
+  set_goals (new_goals ++ later_goals)
+
+end interactive
+end tactic
+
+open_locale big_operators
+
+/-- Filtering to the range of a function is equivalent to taking the image by it. -/
+@[simp] lemma finset.filter_mem_range_univ {α β : Sort*}
+  [fintype α] [fintype β] [decidable_eq β]
+  (f : α → β) {_ : decidable_pred (λ (b : β), b ∈ set.range f)} :
+  finset.filter (λ (b : β), b ∈ set.range f) finset.univ = finset.image f finset.univ :=
+by { ext, simp }
+
+@[simp, to_additive]
+lemma finset.prod_filter_univ_exists_eq {α β γ : Type*}
+  [fintype α] [fintype β] [comm_monoid γ]
+  (f : α → β) (h : function.injective f) {_ : decidable_pred (λ (b : β), ∃ a, f a = b)} (g : β → γ) :
+  ∏ x in finset.filter (λ (b : β), ∃ a, f a = b) finset.univ, g x = ∏ a, g (f a) :=
+begin
+  classical,
+  rw ←finset.prod_image (λ x _ y _ (z : f x = f y), h z),
+  refine finset.prod_congr (finset.filter_mem_range_univ f) (λ _ _, rfl),
+end
+
+lemma finset.sum_filter_univ_mem_monoid_hom_range {α β γ : Type*}
+  [fintype α] [fintype β] [group α] [group β] [add_comm_monoid γ]
+  (f : α →* β) (h : function.injective f) {_ : decidable_pred (λ (b : β), b ∈ f.range)} (g : β → γ) :
+  ∑ x in finset.filter (λ (b : β), b ∈ f.range) finset.univ, g x = ∑ a, g (f a) :=
+finset.sum_filter_univ_exists_eq f h g
```

Comment:
[Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there.20code.20for.20X.3F/topic/Summing.20over.20.60monoid_hom.2Erange.60/near/222093999)